### PR TITLE
Add more documentation to 'ble.scan'.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: lib/ble/ble.toit [hunk 1]
```diff
@@ -187,6 +187,20 @@ Since Toit only supports BLE, this flag should always be set.
 */
 BLE-ADVERTISE-FLAGS-BREDR-UNSUPPORTED  ::= 0x04
 
+/**
+A device that is a controller.
+
+This device is both a LE and a BR/EDR controller.
```
Comment (kasperl) [inline lib/ble/ble.toit:193]:
```suggestion
This device is both an LE and a BR/EDR controller.
```

File: lib/ble/ble.toit [hunk 2]
```diff
@@ -187,6 +187,20 @@ Since Toit only supports BLE, this flag should always be set.
...
+*/
+BLE-ADVERTISE-FLAGS-LE-BREDR-CONTROLLER  ::= 0x08
+
+/**
+A device that is a host.
+
+This device is both a LE and a BR/EDR host.
```
Comment (kasperl) [inline lib/ble/ble.toit:200]:
```suggestion
This device is both an LE and a BR/EDR host.
```

File: lib/ble/ble.toit [hunk 3]
```diff
@@ -713,6 +727,69 @@ class DataBlock:
...
+  operator== other/any:
+    if other is not DataBlock: return false
+    other-block := other as DataBlock
+    return type == other-block.type and data == other-block.data
+
+  stringify --add-prefix/bool=true -> string:
+    prefix := add-prefix ? "DataBlock: " : ""
+    if type == TYPE-RAW: return "$(prefix)raw=$data"
+    if type == TYPE-FLAGS:
+      flags-string := ""
```
Comment (kasperl) [inline lib/ble/ble.toit:742]: I think I'd build this up as a list instead and then join.

File: lib/ble/ble.toit [hunk 4]
```diff
@@ -713,6 +727,69 @@ class DataBlock:
...
+      return "$(prefix)name-complete=$name"
+    if type == TYPE-TX-POWER-LEVEL:
+      return "$(prefix)tx-power-level=$tx-power-level"
+    if type == TYPE-SERVICE-DATA-16 or type == TYPE-SERVICE-DATA-32 or type == TYPE-SERVICE-DATA-128:
+      bit-size/int := ?
+      if type == TYPE-SERVICE-DATA-16: bit-size = 16
+      else if type == TYPE-SERVICE-DATA-32: bit-size = 32
+      else if type == TYPE-SERVICE-DATA-128: bit-size = 128
+      else: unreachable
+      data-string := ""
```
Comment (kasperl) [inline lib/ble/ble.toit:780]: I think I'd build this up as a list instead and then join.

File: lib/ble/ble.toit [hunk 5]
```diff
@@ -713,6 +727,69 @@ class DataBlock:
...
+      if (flags & BLE-ADVERTISE-FLAGS-GENERAL-DISCOVERY) != 0:
+        flags-string += "LE General Discoverable Mode, "
+      if (flags & BLE-ADVERTISE-FLAGS-BREDR-UNSUPPORTED) != 0:
+        flags-string += "BR/EDR Not Supported, "
+      if (flags & BLE-ADVERTISE-FLAGS-LE-BREDR-CONTROLLER) != 0:
+        flags-string += "LE and BR/EDR Controller, "
+      if (flags & BLE-ADVERTISE-FLAGS-LE-BREDR-HOST) != 0:
+        flags-string += "LE and BR/EDR Host, "
+      if flags-string.is-empty: return "$(prefix)flags=$(%x flags)"
+      flags-string = flags-string[.. flags-string.size - 2]
```
Comment (kasperl) [inline lib/ble/ble.toit:754]: Are you dropping the last ", " like this? Hmm. At least it deserves a comment, but using join would be a lot nicer.

File: lib/ble/ble.toit [hunk 6]
```diff
@@ -713,6 +727,69 @@ class DataBlock:
...
+    if type == TYPE-SERVICE-DATA-16 or type == TYPE-SERVICE-DATA-32 or type == TYPE-SERVICE-DATA-128:
+      bit-size/int := ?
+      if type == TYPE-SERVICE-DATA-16: bit-size = 16
+      else if type == TYPE-SERVICE-DATA-32: bit-size = 32
+      else if type == TYPE-SERVICE-DATA-128: bit-size = 128
+      else: unreachable
+      data-string := ""
+      service-data: | uuid/BleUuid data/io.Data |
+        data-string += "$uuid=$data, "
+      data-string = data-string[.. data-string.size - 2]
```
Comment (kasperl) [inline lib/ble/ble.toit:783]: Are you dropping the last ", " like this? Hmm. At least it deserves a comment, but using join would be a lot nicer.

File: lib/ble/ble.toit [hunk 7]
```diff
@@ -713,6 +727,69 @@ class DataBlock:
...
+      return "$(prefix)service-data-$bit-size=($data-string)"
+    if type == TYPE-MANUFACTURER-SPECIFIC:
+      manufacturer-specific: | company-id/ByteArray manufacturer-specific/io.Data |
+        company-id-string := "$(%02x company-id[1])$(%02x company-id[0])"
+        return "$(prefix)manufacturer-specific=$company-id-string-$manufacturer-specific"
+      unreachable
+    return "$(prefix)type=$type, data=$data"
+    if type == TYPE-SERVICE-DATA-32:
```
Comment (kasperl) [inline lib/ble/ble.toit:791]: wat

File: lib/ble/remote.toit [hunk 8]
```diff
@@ -70,6 +70,25 @@ class Central extends Resource_:
...
+
+  A simple way to merge advertisement data is to concatenate the $DataBlock entries of
+    the advertisements.
+
+  ```
+  discovered-blocks := {:}
+  central.scan --duration=(Duration --s=2) --active: | device/RemoteScannedDevice |
+    blocks := discovered-blocks.get device.identifier --init=: {}
+    blocks.add-all device.data.data-blocks
+  // A map from identifier to the discovered advertisements.
```
Comment (kasperl) [inline lib/ble/remote.toit:88]:
```suggestion
  // Construct a map from identifier to the discovered advertisements.
```

File: examples/ble/scan-active.toit [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  adapter := ble.Adapter
+  central := adapter.central
+
+  // An active scan may need the responses to be merged.
+  discovered-blocks := {:}
+  central.scan --duration=(Duration --s=2) --active: | device/ble.RemoteScannedDevice |
+    blocks := discovered-blocks.get device.identifier --init=: {}
+    blocks.add-all device.data.data-blocks
+
+  // A map from identifier to the discovered advertisements.
```
Comment (kasperl) [inline examples/ble/scan-active.toit:20]:
```suggestion
  // Construct a map from identifier to the discovered advertisements.
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: lib/ble/ble.toit (hunk 7 above)
Comment (floitsch) [reply lib/ble/ble.toit]: done.

File: lib/ble/ble.toit (hunk 6 above)
Comment (floitsch) [reply lib/ble/ble.toit]: done.

File: lib/ble/ble.toit (hunk 4 above)
Comment (floitsch) [reply lib/ble/ble.toit]: done.

File: lib/ble/ble.toit (hunk 5 above)
Comment (floitsch) [reply lib/ble/ble.toit]: done.

File: lib/ble/ble.toit (hunk 3 above)
Comment (floitsch) [reply lib/ble/ble.toit]: done.

**Pull request merged**